File: legacy/python-backend/app/tasks/position_monitor.py
import asyncio
import traceback

from app.portfolio.portfolio_manager import (
    PortfolioManager
)

from app.core.database import (
    AsyncSessionLocal
)

from app.services.position.position_service import (
    PositionService
)

from app.core.logger import logger


async def main():

    while True:

        async with AsyncSessionLocal() as db:
            position_service = PositionService(db)
            positions = await position_service.get_all_positions()
            await PortfolioManager.load_positions(positions)

        for symbol, position in (
            PortfolioManager.positions.items()
        ):

            logger.info(
                f"Monitoring Position: "
                f"{symbol} "
                f"SL={position.get('stop_loss')} "
                f"TP={position.get('take_profit')}"
            )

        await asyncio.sleep(10)


async def run_with_recovery():
    while True:
        try:
            await main()
        except Exception as e:
            logger.error(f"Position Monitor crashed: {e}")
            logger.error(traceback.format_exc())
            logger.info("Restarting Position Monitor in 5 seconds...")
            await asyncio.sleep(5)


if __name__ == "__main__":

    asyncio.run(run_with_recovery())

File: backend/src/workers/positionMonitor.ts
import { PositionService } from "../services/positionService";
import { PortfolioManager } from "../trading/portfolio/portfolioManager";
import { logger } from "../core/logger";

async function main() {
  while (true) {
    try {
      const positions = await PositionService.getAllPositions();
      await PortfolioManager.loadPositions(positions);

      for (const [symbol, position] of Object.entries(PortfolioManager.positions)) {
        logger.info(
          `Monitoring Position: ${symbol} SL=${position.stop_loss} TP=${position.take_profit}`
        );
      }
    } catch (e: any) {
      logger.error(`Error in Position Monitor cycle: ${e.message}`);
    }

    await new Promise((resolve) => setTimeout(resolve, 10000));
  }
}

async function runWithRecovery() {
  while (true) {
    try {
      await main();
    } catch (e: any) {
      logger.error(`Position Monitor crashed: ${e.message}`);
      if (e.stack) logger.error(e.stack);
      logger.info("Restarting Position Monitor in 5 seconds...");
      await new Promise((resolve) => setTimeout(resolve, 5000));
    }
  }
}

runWithRecovery();
